File: src/ainfra/doctor.py
"""Read-only local readiness checks."""

from __future__ import annotations

import shutil
import subprocess
import sys
import tomllib
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any

from ainfra.contracts import repository_root, validate_path
from ainfra.policy import validate_policy
from ainfra.template import discover_template


@dataclass(frozen=True, slots=True)
class Check:
    """One machine-readable doctor result."""

    id: str
    status: str
    found: str | None
    required: str
    remediation: str


COMMANDS = {
    "uv": ["uv", "--version"],
    "tofu": ["tofu", "version"],
    "ansible-playbook": ["ansible-playbook", "--version"],
    "checkov": ["checkov", "--version"],
    "gitleaks": ["gitleaks", "version"],
}


def run_doctor(input_path: Path | None = None) -> list[Check]:
    """Return readiness checks without installing or changing anything."""

    checks = [
        Check(
            "python",
            "pass" if sys.version_info[:2] == (3, 12) else "fail",
            ".".join(str(part) for part in sys.version_info[:3]),
            "Python 3.12",
            "Run through `uv run` with the committed .python-version.",
        )
    ]
    root = repository_root()
    tool_pins = _load_tool_pins(root / "tools.lock")
    for check_id, argv in COMMANDS.items():
        pin = tool_pins[check_id]
        required = f"{pin['policy']} {pin['version']}"
        executable = shutil.which(argv[0])
        if executable is None:
            checks.append(
                Check(
                    check_id,
                    "fail",
                    None,
                    required,
                    f"Install {argv[0]} and make it available on PATH.",
                )
            )
            continue
        result = subprocess.run(
            argv,
            check=False,
            capture_output=True,
            text=True,
            timeout=10,
        )
        output = (result.stdout or result.stderr).splitlines()
        first_line = output[0] if output else executable
        found_version = _extract_version(first_line)
        version_ok = (
            result.returncode == 0
            and found_version is not None
            and _version_matches(
                found_version,
                str(pin["version"]),
                str(pin["policy"]),
            )
        )
        checks.append(
            Check(
                check_id,
                "pass" if version_ok else "fail",
                found_version or first_line,
                required,
                f"Install {argv[0]} at {required}.",
            )
        )

    checks.extend(
        [
            _file_check(root / "uv.lock", "uv-lock"),
            _file_check(root / "tools.lock", "tool-lock"),
            _file_check(root / ".gitignore", "gitignore"),
            _python_pin_check(root / "pyproject.toml"),
            _template_check(),
            _backend_policy_check(input_path),
            Check(
                "github-workflows",
                (
                    "fail"
                    if (root / ".github" / "workflows").exists()
                    else "pass"
                ),
                None,
                "no GitHub workflows",
                "Remove .github/workflows; all automation is local.",
            ),
        ]
    )
    return checks


def serialized_checks(
    checks: list[Check] | None = None,
) -> list[dict[str, str | None]]:
    """Serialize doctor output without exposing local environment details."""

    return [asdict(check) for check in checks or run_doctor()]


def _file_check(path: Path, check_id: str) -> Check:
    return Check(
        check_id,
        "pass" if path.is_file() else "fail",
        str(path) if path.is_file() else None,
        f"{path.name} present",
        f"Restore the committed {path.name} file.",
    )


def _load_tool_pins(path: Path) -> dict[str, dict[str, Any]]:
    with path.open("rb") as handle:
        document = tomllib.load(handle)
    tools = document.get("tools")
    if not isinstance(tools, dict):
        raise ValueError("tools.lock must contain [tools.*] entries")
    return tools


def _extract_version(output: str) -> str | None:
    import re

    match = re.search(r"(?<!\d)(\d+\.\d+(?:\.\d+)?)", output)
    return match.group(1) if match else None


def _version_tuple(value: str) -> tuple[int, ...]:
    return tuple(int(part) for part in value.split("."))


def _version_matches(found: str, required: str, policy: str) -> bool:
    if policy == "exact":
        return _version_tuple(found) == _version_tuple(required)
    if policy == "minimum":
        return _version_tuple(found) >= _version_tuple(required)
    return False


def _python_pin_check(path: Path) -> Check:
    with path.open("rb") as handle:
        document = tomllib.load(handle)
    project = document["project"]
    dependencies = project.get("dependencies", [])
    dev_dependencies = document.get("dependency-groups", {}).get("dev", [])
    all_dependencies = [*dependencies, *dev_dependencies]
    pinned = all("==" in dependency for dependency in all_dependencies)
    return Check(
        "python-pins",
        "pass" if pinned else "fail",
        ", ".join(all_dependencies),
        "all Python dependencies exactly pinned",
        "Pin every runtime and development dependency with ==.",
    )


def _template_check() -> Check:
    try:
        template = discover_template("hetzner-kubernetes-baseline")
    except Exception as exc:
        return Check(
            "template-contract",
            "fail",
            str(exc),
            "reference template resolves safely",
            "Repair the reference manifest and its contained paths.",
        )
    return Check(
        "template-contract",
        "pass",
        str(template.manifest_path),
        "reference template resolves safely",
        "Repair the reference manifest and its contained paths.",
    )


def _backend_policy_check(input_path: Path | None = None) -> Check:
    root = repository_root()
    path = input_path or (
        root
        / "templates"
        / "hetzner-kubernetes-baseline"
        / "inputs"
        / "example.input.yaml"
    )
    try:
        document = validate_path(path)
        validate_policy(
            document,
            source=path,
            template_name="hetzner-kubernetes-baseline",
        )
    except Exception as exc:
        return Check(
            "backend-policy",
            "fail",
            str(exc),
            "example input satisfies state policy",
            "Repair the example state mode and backend capabilities.",
        )
    return Check(
        "backend-policy",
        "pass",
        str(path),
        "example input satisfies state policy",
        "Repair the example state mode and backend capabilities.",
    )

File: src/doctor.rs
//! Read-only local readiness checks.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::process::Command;

use regex::Regex;
use serde::Serialize;

use crate::contracts::validate_path;
use crate::policy::validate_policy;
use crate::template::discover_builtin;

/// One stable, machine-readable readiness result.
#[derive(Clone, Debug, Serialize)]
pub struct Check {
    /// Stable check identifier.
    pub id: String,
    /// `pass` or `fail`.
    pub status: String,
    /// Detected value, when available.
    pub found: Option<String>,
    /// Required state.
    pub required: String,
    /// Operator remediation.
    pub remediation: String,
}

impl Check {
    fn new(
        id: &str,
        passed: bool,
        found: Option<String>,
        required: impl Into<String>,
        remediation: impl Into<String>,
    ) -> Self {
        Self {
            id: id.to_owned(),
            status: if passed { "pass" } else { "fail" }.to_owned(),
            found,
            required: required.into(),
            remediation: remediation.into(),
        }
    }
}

/// Run the compatibility doctor against an explicit project root.
#[must_use]
pub fn run_doctor(root: &Path, input: Option<&Path>) -> Vec<Check> {
    let mut checks = Vec::new();
    let pins = load_tool_pins(&root.join("tools.lock"));
    for (id, argv) in [
        ("tofu", &["tofu", "version"][..]),
        ("ansible-playbook", &["ansible-playbook", "--version"][..]),
        ("checkov", &["checkov", "--version"][..]),
        ("gitleaks", &["gitleaks", "version"][..]),
    ] {
        checks.push(command_check(id, argv, pins.get(id)));
    }
    for (name, id) in [("tools.lock", "tool-lock"), (".gitignore", "gitignore")] {
        checks.push(file_check(&root.join(name), id));
    }
    checks.push(template_check());
    checks.push(backend_policy_check(root, input));
    checks.push(Check::new(
        "github-workflows",
        !root.join(".github/workflows").exists(),
        None,
        "no GitHub workflows",
        "Remove .github/workflows; all automation is local.",
    ));
    checks
}

/// Run installed-product readiness checks for one validated project input.
#[must_use]
pub fn run_project_doctor(root: &Path, input: &Path) -> Vec<Check> {
    let tofu = ("1.10.0".to_owned(), "minimum".to_owned());
    let ansible = ("2.16.0".to_owned(), "minimum".to_owned());
    vec![
        command_check("tofu", &["tofu", "version"], Some(&tofu)),
        command_check(
            "ansible-playbook",
            &["ansible-playbook", "--version"],
            Some(&ansible),
        ),
        template_check(),
        backend_policy_check(root, Some(input)),
    ]
}

fn command_check(id: &str, argv: &[&str], pin: Option<&(String, String)>) -> Check {
    let (required_version, policy) = pin
        .cloned()
        .unwrap_or_else(|| ("unknown".to_owned(), "exact".to_owned()));
    let required = format!("{policy} {required_version}");
    let line = command_first_line(argv);
    let found = line.as_deref().and_then(extract_version).map(str::to_owned);
    let passed = found
        .as_deref()
        .is_some_and(|value| version_matches(value, &required_version, &policy));
    Check::new(
        id,
        passed,
        found.or(line),
        required.clone(),
        format!("Install {} at {required}.", argv[0]),
    )
}

fn command_first_line(argv: &[&str]) -> Option<String> {
    let output = Command::new(argv.first()?).args(&argv[1..]).output().ok()?;
    if !output.status.success() {
        return None;
    }
    let text = if output.stdout.is_empty() {
        &output.stderr
    } else {
        &output.stdout
    };
    String::from_utf8_lossy(text)
        .lines()
        .next()
        .map(str::to_owned)
}

fn extract_version(value: &str) -> Option<&str> {
    let pattern = Regex::new(r"\d+\.\d+(?:\.\d+)?").ok()?;
    pattern.find(value).map(|found| found.as_str())
}

fn version_matches(found: &str, required: &str, policy: &str) -> bool {
    let parts = |value: &str| {
        value
            .split('.')
            .map(str::parse::<u64>)
            .collect::<Result<Vec<_>, _>>()
            .ok()
    };
    match (parts(found), parts(required), policy) {
        (Some(found), Some(required), "exact") => found == required,
        (Some(found), Some(required), "minimum") => found >= required,
        _ => false,
    }
}

fn load_tool_pins(path: &Path) -> BTreeMap<String, (String, String)> {
    let Ok(document) = std::fs::read_to_string(path) else {
        return BTreeMap::new();
    };
    let mut result = BTreeMap::new();
    let mut current = None;
    let mut version = None;
    let mut policy = None;
    for line in document.lines().map(str::trim) {
        if let Some(name) = line
            .strip_prefix("[tools.")
            .and_then(|value| value.strip_suffix(']'))
        {
            if let (Some(name), Some(version), Some(policy)) =
                (current.take(), version.take(), policy.take())
            {
                result.insert(name, (version, policy));
            }
            current = Some(name.to_owned());
        } else if let Some(value) = assignment(line, "version") {
            version = Some(value);
        } else if let Some(value) = assignment(line, "policy") {
            policy = Some(value);
        }
    }
    if let (Some(name), Some(version), Some(policy)) = (current, version, policy) {
        result.insert(name, (version, policy));
    }
    result
}

fn assignment(line: &str, key: &str) -> Option<String> {
    line.strip_prefix(key)?
        .trim_start()
        .strip_prefix('=')?
        .trim()
        .strip_prefix('"')?
        .strip_suffix('"')
        .map(str::to_owned)
}

fn file_check(path: &Path, id: &str) -> Check {
    Check::new(
        id,
        path.is_file(),
        path.is_file().then(|| path.display().to_string()),
        format!(
            "{} present",
            path.file_name().unwrap_or_default().to_string_lossy()
        ),
        format!(
            "Restore the committed {} file.",
            path.file_name().unwrap_or_default().to_string_lossy()
        ),
    )
}

fn template_check() -> Check {
    match discover_builtin("hetzner-kubernetes-baseline") {
        Ok(template) => Check::new(
            "template-contract",
            true,
            Some(template.manifest_path.to_owned()),
            "reference template resolves safely",
            "Repair the reference manifest and its contained paths.",
        ),
        Err(error) => Check::new(
            "template-contract",
            false,
            Some(error.to_string()),
            "reference template resolves safely",
            "Repair the reference manifest and its contained paths.",
        ),
    }
}

fn backend_policy_check(root: &Path, input: Option<&Path>) -> Check {
    let default = root.join("templates/hetzner-kubernetes-baseline/inputs/example.input.yaml");
    let path: PathBuf = input.map_or(default, Path::to_path_buf);
    let result = validate_path(&path).and_then(|document| {
        validate_policy(&document, &path, Some("hetzner-kubernetes-baseline"), root)
    });
    match result {
        Ok(()) => Check::new(
            "backend-policy",
            true,
            Some(path.display().to_string()),
            "example input satisfies state policy",
            "Repair the example state mode and backend capabilities.",
        ),
        Err(error) => Check::new(
            "backend-policy",
            false,
            Some(error.to_string()),
            "example input satisfies state policy",
            "Repair the example state mode and backend capabilities.",
        ),
    }
}
